### ISSUE TITLE
Constant offsets causing loud pops

### ISSUE THREAD

**Issue by Schroedingers-Cat** [ValveSoftware/steam-audio#300]
**System Information**
- Steam Audio version: 4.1.4 - 4.5.1 
  (I didn't test any versions below 4.1.4 but I used v2.0.20 in another project where this never happened)
- Unity version: 2022.3.17f1
- Operating System and version: Windows 10, Linux (Kernel 6.6/Manjaro/Gnome 45.3/Wayland)
- CPU architecture: x86-64
- RME HDSPe MADI configured as 5.1 output on Windows, configured as 5.1 PA sink on Linux
- CPU: AMD Ryzen 9 3950X, GPU: AMD Radeon RX 6900 XT

**Issue Description**
Using Steam Audio, it's possible to produce sudden constant offsets and pops. No DSP should ever do that as it can damage peoples hearing and their gear (and their trust).

**Steps To Reproduce**
1. Download the [repro case](https://owncloud.gwdg.de/index.php/s/rwWs2sbdrqPRPS2) and extract it into a new directory
2. Open it with Unity 2022.3.17f1
3. Load the included scene `Scenes/Bug-SteamAudioOffsetPops/SteamAudio-OffsetPops`
4. Unplug all audio speakers or turn down the volume! Just monitor what is happening using audio engineering monitoring tools or use SteamAudio's internal debugging options. There is no point in listening to pops caused by offsets.
5. Enter Play Mode
6. Keep scene and game view tab open so you can understand what is going on
7. Use WASD to walk 3 steps on the grid in front of you (so the `GridBasedFPSController` lands on the red marked area, perfectly lined up to the Audio Source `Audio Water Drips`)
8. Use the arrow keys to turn the camera towards the Audio Source `Audio Water Drips` (so that you're looking at the GameObject `Audio Water Drops`)
9. Note the constant offset on the left channel (causing a loud pop on your speakers)

![grafik](https://github.com/ValveSoftware/steam-audio/assets/6382400/49d9f90b-3d51-4653-9fa0-8f77fd28f130)

Further notes:
- doesn't happen when enabling `Direct Binaural` on the Audio Sources
- this is just one specific pop issue that I have been able to isolate
  - in a slightly more complex project, the constant offsets happen in various situations 
  - with a lot more output channels affected than just the left speaker 
  - with more platforms affected (macOS Ventura/ARM64 using the MacBook's built-in stereo speakers)

**Comment by lakulish** [comment]
@Schroedingers-Cat I followed the steps above (thanks for the detailed repro steps!), but was unfortunately unable to reproduce the issue. Can you try the following:

* In the main menu, click Steam Audio > Settings.
* Under Advanced Settings, check the box next to Enable Validation.

This should hopefully result in some console messages when the constant offset occurs. If so, can you share the console messages? That will help us narrow down the issue.

**Comment by Schroedingers-Cat** [comment]
@lakulish thanks a lot for looking into this!  

Setting `Enable Validation` to true doesn't help with the logging, unfortunately. The only message is the default HRTF message when entering play mode:
```
Loaded HRTF: default.
UnityEngine.Debug:Log (object)
SteamAudio.HRTF:.ctor (SteamAudio.Context,SteamAudio.AudioSettings,string,byte[],single,SteamAudio.HRTFNormType) (at Assets/Plugins/SteamAudio/Scripts/Runtime/HRTF.cs:52)
SteamAudio.SteamAudioManager:OnApplicationStart (SteamAudio.ManagerInitReason) (at Assets/Plugins/SteamAudio/Scripts/Runtime/SteamAudioManager.cs:298)
SteamAudio.SteamAudioManager:Initialize (SteamAudio.ManagerInitReason) (at Assets/Plugins/SteamAudio/Scripts/Runtime/SteamAudioManager.cs:637)
SteamAudio.SteamAudioManager:AutoInitialize () (at Assets/Plugins/SteamAudio/Scripts/Runtime/SteamAudioManager.cs:1508)
```

I noticed that the issue happens with the following `Default Speaker Mode` settings in Unity's project settings:
- Surround 7.1
- Surround 5.1
- Quad

Setting `Stereo` as `Default Speaker Mode` makes the issue disappear. Maybe you're trying to reproduce this with a stereo output? You can check that via a C# script using `AudioSettings.GetConfiguration().speakerMode`.

**Comment by Schroedingers-Cat** [comment]
@lakulish I created another repro case that causes pops on my MacBook (ARM/M1 Pro/Stereo) as well as Win10 (x86_64/AMD CPU,GPU/7.1).

1. Download and extract [the project](https://owncloud.gwdg.de/index.php/s/ws04lpWhcGukz9t)
2. Open with Unity v2022.3.17f1 and load scene `Assets\Scenes\SteamAudio-PopsInStereo.unity`
3. Open Steam Audio Settings
4. Enter play mode
5. Immediately after the editor becomes responsive change the `Real Time Duration` value from `1` to something else like `2` using numeric keyboard input (I didn`t use the mouse to slide)
6. Pops caused by constant offsets will be all over the place (if that didn't work, continue with the following steps)
7. Stop play mode
8. Reset Steam Audio Settings `Real Time Duration` to `1`
9. Go to four and repeat until step 6 works

For five tries, I get the constant offset at least for once.  

While this might not be a common use case, it might help us get to the bottom of these constant offsets.

The validation setting is already enabled with this project but I don't get it to log anything of interest.

**Comment by lakulish** [comment]
@Schroedingers-Cat This turned out to be two separate issues.

The first one, that can't be reproduced when using stereo output, was a bug in quad/5.1/7.1 panning that only occurs when the source is exactly to your left. I was unable to reproduce this earlier because even though I had set my output format in Unity to 7.1, my audio driver was forcing everything to stereo. I changed that, and was able to reproduce the issue. A fix will be included in the next release.

The second issue was due to Steam Audio not correctly handling the case where the reflections duration is increased after entering play mode. That control is only intended to _decrease_ the value to reduce CPU usage if needed. I was able to reproduce this issue as well, and a fix will be included in the next release.

**Comment by Schroedingers-Cat** [comment]
@lakulish thank you so much for fixing this!  

If these were separate issues, then it could be that my original project still produces constant offsets caused by other circumstances. I wonder what would be the best way to tackle this:
- Should I wait for the next release and report back to this issue?
- Should I send you the actual scene (+ code + audio) via email?
- Could there be similar problems with surround setups causing these offsets in other contexts?
- Do you see a way to prevent these constant offsets in a more generic way?

**Comment by lakulish** [comment]
The above two issues have been fixed in v4.5.2: https://github.com/ValveSoftware/steam-audio/releases/tag/v4.5.2. I'm closing this issue, but feel free to create new issues if you observe similar behavior in other contexts. Thanks!
